bindings/src/main/java/io/pravega/storage/gcp/GCPChunkStorage.java

```diff
@@ -0,0 +1,249 @@
+ */
+@Slf4j
+public class GCPChunkStorage extends BaseChunkStorage {
+
+    /**
+     * Error code for not found in google cloud storage
+     */
+    private static final int FILE_NOT_FOUND = 404;
+
+    //region members
+    /**
+     * GCPStorageConfig contains config values.
+     */
+    private final GCPStorageConfig config;
+
+    /**
+     * Storage is provided by GCP to perform cloud operation on google cloud storage.
+     */
+    private final Storage storage;
+
+    //endregion
+
+    //region constructor
+    public GCPChunkStorage(Storage storage, GCPStorageConfig config, Executor executor, boolean shouldCloseClient) {
+        super(executor);
+        this.config = Preconditions.checkNotNull(config, "config");
+        this.storage = Preconditions.checkNotNull(storage, "client");
+    }
+    //endregion
+
+    //region capabilities
+
+    @Override
+    public boolean supportsConcat() {
+        return false;
+    }
+
+    @Override
+    public boolean supportsAppend() {
+        return false;
+    }
+
+    @Override
+    public boolean supportsTruncation() {
+        return false;
+    }
+
+    //endregion
+
+    //region implementation
+
+    @Override
+    protected ChunkHandle doOpenRead(String chunkName) throws ChunkStorageException {
+        if (!checkExists(chunkName)) {
+            throw new ChunkNotFoundException(chunkName, "doOpenRead");
+        }
+        return ChunkHandle.readHandle(chunkName);
+    }
+
+    @Override
+    protected ChunkHandle doOpenWrite(String chunkName) throws ChunkStorageException {
+        if (!checkExists(chunkName)) {
+            throw new ChunkNotFoundException(chunkName, "doOpenWrite");
+        }
+        return new ChunkHandle(chunkName, false);
+    }
+
+    @Override
+    protected int doRead(ChunkHandle handle, long fromOffset, int length, byte[] buffer, int bufferOffset) throws ChunkStorageException {
+        try (ReadChannel readChannel = this.storage.reader(this.config.getBucket(), getObjectPath(handle.getChunkName()))) {
+            try {
+                readChannel.seek(fromOffset);
+                readChannel.limit(fromOffset + length);
+                int val = readChannel.read(ByteBuffer.wrap(buffer, bufferOffset, length));
+                if (val == -1) {
+                    ChunkInfo info = doGetInfo(handle.getChunkName());
+                    if (fromOffset >= info.getLength() || (length + fromOffset) >= info.getLength()) {
+                        throw new IllegalArgumentException(handle.getChunkName());
+                    }
+                }
+                return val;
+            } catch (Exception e) {
+                throw convertException(handle.getChunkName(), "doRead", e);
+            }
+        }
+    }
+
+    @Override
+    protected int doWrite(ChunkHandle handle, long offset, int length, InputStream data) {
+        throw new UnsupportedOperationException("GCPChunkStorage does not support writing to already existing objects.");
```

Comment:
just to get enough coverage please add something like following tests  `ChunkStorageTests`

```
@Test 
public void testUnsupported() {
  if (!chunkStorage.supportsAppend()) {
      AssertExtensions.assertFutureThrows(
                " write should throw UnsupportedOperationException.",
chunkStorage.write(...),
                ex -> (ex.getCause() instanceof UnsupportedOperationException || ex instanceof UnsupportedOperationException);
  }

  if (!chunkStorage.supportsConcat()) {
         AssertExtensions.assertFutureThrows(
                " concat should throw UnsupportedOperationException.",
chunkStorage.concat(...),
                ex -> (ex.getCause() instanceof UnsupportedOperationException || ex instanceof UnsupportedOperationException);
  }
}
```